### PR TITLE
Compile references and dereferences.

### PR DESCRIPTION
**Pull request by ptersilie**
I've left a few `todo`'s in the code to try and keep this PR small. We won't be needing them right now so we can implement them at a later stage.

### CODE REVIEW

File: ykcompile/src/lib.rs [hunk 1]
```diff
@@ -251,11 +257,77 @@ impl<TT> TraceCompiler<TT> {
                     ; mov [rdi + aoff], rax
                 );
             }
+            (Location::Register(reg), Location::Deref(boxed)) => match *boxed {
+                Location::Stack(off) => {
+                    dynasm!(self.asm
+                        ; mov Rq(reg), [rbp - off]
```
Comment (vext01) [inline ykcompile/src/lib.rs:263]: Do I see it right that we dereference two things here?

Comment (vext01) [reply ykcompile/src/lib.rs:263]: Once to get the pointer off the stack, again to dereference the pointer?

File: ykcompile/src/lib.rs [hunk 2]
```diff
@@ -251,11 +257,77 @@ impl<TT> TraceCompiler<TT> {
...
         }
         Ok(())
     }
 
+    fn mov_place_ref(&mut self, p1: &Place, p2: &Place) -> Result<(), CompileError> {
+        let lloc = self.place_to_location(p1)?;
+        // We can only reference Locals living on the stack. So move it there if it doesn't.
+        let rloc = match self.place_to_location(p2)? {
+            Location::Register(reg) => {
+                let offset = self.stack_builder.alloc(8, 8) as i32;
```
Comment (vext01) [inline ykcompile/src/lib.rs:300]: (8, 8) here is the size and align of a pointer? I wonder if it should be a constant?

File: ykcompile/src/lib.rs (hunk 1 above)
Comment (ptersilie) [reply ykcompile/src/lib.rs:263]: Yes and no. We have a reference/pointer on the stack (in `[rbp-off]`). So we first move that pointer into a register (first instruction), and then dereference the pointer to get its value (second instruction).

File: ykcompile/src/lib.rs [hunk 3]
```diff
@@ -1331,4 +1427,90 @@ mod tests {
...
+
+    #[test]
+    fn test_ref_deref() {
+        let mut inputs = trace_inputs((0,));
+        let th = start_tracing(Some(TracingKind::HardwareTracing));
+        inputs.0 = ref_deref();
+        let sir_trace = th.stop_tracing().unwrap();
+        let tir_trace = TirTrace::new(&*sir_trace).unwrap();
+        let ct = TraceCompiler::<&(u64,)>::compile(tir_trace);
+        let mut args = (0,);
```
Comment (vext01) [inline ykcompile/src/lib.rs:1447]: Let's get into the habit of trying different args in each test case. I think it'll reduce "works by luck" test passes.

File: ykcompile/src/lib.rs (hunk 2 above)
Comment (ptersilie) [reply ykcompile/src/lib.rs:300]: Dunno, haven't really been thinking about the type stuff yet, so just copy-pasted from above. We can fix that when we do the type changes?

File: ykcompile/src/lib.rs [hunk 4]
```diff
@@ -106,6 +106,7 @@ enum Location {
     Register(u8),
     Stack(i32),
     Arg(i32),
+    Deref(Box<Location>),
```
Comment (vext01) [inline ykcompile/src/lib.rs:109]: Perhaps not for this PR, but IIRC we agreed that all of the projections would fall under one "Projection" variant?

I'm also still wondering if `Stack(offset)` should be `Mem(reg, offset)` to handle things like heap allocs or offsets from temporary registers, but again, possibly for another PR.

File: ykcompile/src/lib.rs (hunk 1 above)
Comment (vext01) [reply ykcompile/src/lib.rs:263]: ok

File: ykcompile/src/lib.rs (hunk 3 above)
Comment (ptersilie) [reply ykcompile/src/lib.rs:1447]: It depends what you do with the args. If you overwrite it it won't matter, but if the output depends on it, we can't just randomly choose an input value.

File: ykcompile/src/lib.rs (hunk 2 above)
Comment (vext01) [reply ykcompile/src/lib.rs:300]: ok

File: ykcompile/src/lib.rs (hunk 3 above)
Comment (vext01) [reply ykcompile/src/lib.rs:1447]: Ah, I see. The "input" here is actually only an output.

Comment (ptersilie) [reply ykcompile/src/lib.rs:1447]: Yes.

File: ykcompile/src/lib.rs (hunk 4 above)
Comment (ptersilie) [reply ykcompile/src/lib.rs:109]: I agree. One thing about `Deref`s though is that I think it needs to be recursive, which is why I just made it part of `Location`.

I always assumed that some of these things will change later on. It's hard to say what this stuff needs to look like until we have actual examples (e.g. heap alloc) to refer to. And we can't really solve everything in one go. So we'll have to come back to it in another PR.

Comment (vext01) [reply ykcompile/src/lib.rs:109]: Yep, that's fine.

Let's be on the lookout for variants in Location that can be generalised later.

**Comment by ptersilie** [comment]
If there are no comments for me to address, then I think this is ready for merging.

**Pull request merged**